File: app/src/main/java/com/example/finalexer3grp2/FilmAdapter.java
package com.example.finalexer3grp2;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import coil.Coil;
import coil.request.ImageRequest;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class FilmAdapter extends RecyclerView.Adapter<FilmAdapter.FilmViewHolder> {

    private List<Film> films;
    private final OnItemClickListener onItemClickListener;
    private final OnSelectionChangedListener onSelectionChangedListener;

    private final Set<Integer> selectedIds = new HashSet<>();
    private boolean isSelectionMode = false;

    public interface OnItemClickListener {
        void onItemClick(Film film);
    }

    public interface OnSelectionChangedListener {
        void onSelectionChanged(int count);
    }

    public FilmAdapter(List<Film> films, OnItemClickListener onItemClickListener, OnSelectionChangedListener onSelectionChangedListener) {
        this.films = films;
        this.onItemClickListener = onItemClickListener;
        this.onSelectionChangedListener = onSelectionChangedListener;
    }

    public void clearSelection() {
        selectedIds.clear();
        isSelectionMode = false;
        notifyDataSetChanged();
        onSelectionChangedListener.onSelectionChanged(0);
    }

    public boolean isSelectionMode() {
        return isSelectionMode;
    }

    public Set<Integer> getSelectedIds() {
        return selectedIds;
    }

    public Film getFilmAt(int position) {
        if (position >= 0 && position < films.size()) {
            return films.get(position);
        }
        return null;
    }

    public void toggleSelection(int filmId) {
        if (selectedIds.contains(filmId)) {
            selectedIds.remove(filmId);
        } else {
            selectedIds.add(filmId);
        }

        if (selectedIds.isEmpty()) {
            isSelectionMode = false;
        }

        onSelectionChangedListener.onSelectionChanged(selectedIds.size());
    }

    public static class FilmViewHolder extends RecyclerView.ViewHolder {
        final TextView title;
        final TextView info;
        final ImageView poster;

        public FilmViewHolder(@NonNull View view) {
            super(view);
            title = view.findViewById(R.id.itemTitle);
            info = view.findViewById(R.id.itemYear);
            poster = view.findViewById(R.id.itemPoster);
        }
    }

    @NonNull
    @Override
    public FilmViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_film, parent, false);
        return new FilmViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull FilmViewHolder holder, int position) {

        Film film = films.get(position);
        boolean isSelected = selectedIds.contains(film.getId());

        SimpleDateFormat sdf = new SimpleDateFormat("MMMM dd yyyy", Locale.getDefault());

        String updatedDate = sdf.format(new Date(film.getUpdatedAt()));

        holder.title.setText(film.getTitle());
        holder.info.setText(film.getYear() + " • " + film.getDirector() + "\nUpdated: " + updatedDate);

        // Coil Java usage
        ImageRequest request = new ImageRequest.Builder(holder.poster.getContext())
                .data(film.getPosterUri())
                .crossfade(true)
                .placeholder(R.drawable.ic_launcher_background)
                .error(R.drawable.ic_launcher_background)
                .target(holder.poster)
                .build();
        Coil.imageLoader(holder.poster.getContext()).enqueue(request);

        // color when selected
        if (isSelected) {
            holder.itemView.setForeground(new ColorDrawable(Color.parseColor("#803F51B5")));
            holder.itemView.setAlpha(1.0f);
        } else {
            holder.itemView.setForeground(null);
            holder.itemView.setAlpha(isSelectionMode ? 0.6f : 1.0f);
        }

        holder.itemView.setOnClickListener(v -> {
            if (isSelectionMode) {
                toggleSelection(film.getId());
                notifyItemChanged(position);
            } else {
                onItemClickListener.onItemClick(film);
            }
        });

        holder.itemView.setOnLongClickListener(v -> {
            if (!isSelectionMode) {
                isSelectionMode = true;
                selectedIds.add(film.getId());
                notifyItemChanged(position);
                onSelectionChangedListener.onSelectionChanged(selectedIds.size());
            }
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return films.size();
    }

    public void updateData(List<Film> newFilms) {
        this.films = newFilms;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/finalexer3grp2/FilmAdapter.kt
package com.example.finalexer3grp2

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import coil.load

class FilmAdapter(
    private var films: List<Film>,
    private val onEditClick: (Film) -> Unit,
    private val onSelectionChanged: (Int) -> Unit = {}
) : RecyclerView.Adapter<FilmAdapter.FilmViewHolder>() {

    val selectedIds = mutableSetOf<Int>()
    var isSelectionMode = false

    fun clearSelection() {
        selectedIds.clear()
        isSelectionMode = false
        notifyDataSetChanged()
        onSelectionChanged(0)
    }

    class FilmViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val title: TextView = view.findViewById(R.id.itemTitle)
        val info: TextView = view.findViewById(R.id.itemYear)
        val poster: ImageView = view.findViewById(R.id.itemPoster)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FilmViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_film, parent, false)
        return FilmViewHolder(view)
    }

    override fun onBindViewHolder(holder: FilmViewHolder, position: Int) {
        val film = films[position]
        val isSelected = selectedIds.contains(film.id)

        holder.title.text = film.title
        holder.info.text = "${film.year} • Dir. ${film.director}"
        holder.poster.load(film.posterUri) {
            crossfade(true)
            placeholder(R.drawable.ic_launcher_background)
            error(R.drawable.ic_launcher_background)
        }

        // Visual feedback
        holder.itemView.alpha = if (isSelectionMode && !isSelected) 0.5f else 1.0f
        holder.itemView.isActivated = isSelected // use activated state in your item bg selector

        holder.itemView.setOnLongClickListener {
            if (!isSelectionMode) {
                // Enter selection mode
                isSelectionMode = true
                selectedIds.add(film.id)
                notifyDataSetChanged()
                onSelectionChanged(selectedIds.size)
            } else {
                // Already in selection mode — show popup like before
                val popup = android.widget.PopupMenu(holder.itemView.context, holder.itemView)
                popup.menu.add("Edit")
                popup.setOnMenuItemClickListener { item ->
                    if (item.title == "Edit") onEditClick(film)
                    true
                }
                popup.show()
            }
            true
        }

        holder.itemView.setOnClickListener {
            if (isSelectionMode) {
                if (selectedIds.contains(film.id)) {
                    selectedIds.remove(film.id)
                } else {
                    selectedIds.add(film.id)
                }
                if (selectedIds.isEmpty()) {
                    isSelectionMode = false
                }
                notifyItemChanged(position)
                onSelectionChanged(selectedIds.size)
            }
            // Normal click does nothing here since edit is via long press popup
        }
    }

    override fun getItemCount() = films.size

    fun updateData(newFilms: List<Film>) {
        this.films = newFilms
        notifyDataSetChanged()
    }

    fun toggleSelection(filmId: Int) {
        if (selectedIds.contains(filmId)) {
            selectedIds.remove(filmId)
        } else {
            selectedIds.add(filmId)
        }
        if (selectedIds.isEmpty()) isSelectionMode = false
        onSelectionChanged(selectedIds.size)
    }

    fun getFilmAt(position: Int): Film? {
        return films.getOrNull(position)
    }
}